commons/src/system/daemon.py

```diff
@@ -0,0 +1,190 @@
+#!/usr/bin/env python
+
+#
+# Copyright 2012-2016, the original author or authors.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# You may not use this file except in compliance
+# with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+import sys, os, time, atexit
+from signal import SIGTERM
+
+class Daemon:
+	"""
+	A generic daemon class.
+	
+	Usage: subclass the Daemon class and override the run() method
+	"""
+	def __init__(self, pidfile, stdin='/dev/null', stdout='/dev/null', stderr='/dev/null'):
+		self.stdin = stdin
+		self.stdout = stdout
+		self.stderr = stderr
+		self.pidfile = pidfile
+                #capture path from where script was triggered
+                self.rundir = os.getcwd()
+
+                #If daemon log dir is specified, then log there
+                #TODO: Hack to avoid accepting too many arguments at startup.
+                # Will behave weirdly for relative paths and /dev/null inputs
+                daemon_log_dir = os.getenv('DAEMON_LOG_DIR', None)
+                if daemon_log_dir and not os.path.isabs(self.stdout):
+                    self.stdout = daemon_log_dir + '/' + self.stdout
+                else:
+                    self.stdout = os.path.abspath(self.stdout)
+                if daemon_log_dir and not os.path.isabs(self.stderr):
+                    self.stderr = daemon_log_dir + '/' + self.stderr
+                else:
+                    self.stderr = os.path.abspath(self.stderr)
+	
+	def daemonize(self):
+		"""
+		do the UNIX double-fork magic, see Stevens' "Advanced 
+		Programming in the UNIX Environment" for details (ISBN 0201563177)
+		http://www.erlenstar.demon.co.uk/unix/faq_2.html#SEC16
+		"""
+		try: 
+			pid = os.fork() 
+			if pid > 0:
+				# exit first parent
+				sys.exit(0) 
+		except OSError, e: 
+			sys.stderr.write("fork #1 failed: %d (%s)\n" % (e.errno, e.strerror))
+			sys.exit(1)
+	
+		# decouple from parent environment
+		os.chdir("/") 
+		os.setsid() 
+		os.umask(0) 
+	
+		# do second fork
+		try: 
+			pid = os.fork() 
+			if pid > 0:
+				# exit from second parent
+				sys.exit(0) 
+		except OSError, e: 
+			sys.stderr.write("fork #2 failed: %d (%s)\n" % (e.errno, e.strerror))
+			sys.exit(1) 
+	
+		# redirect standard file descriptors
+		sys.stdout.flush()
+		sys.stderr.flush()
+		si = file(self.stdin, 'r')
+		so = file(self.stdout, 'a+')
```

Comment:
This is not written specifically for Cuppa. Double forking works well. It has been tested thoroughly.